temporalio/worker.py

```diff
@@ -0,0 +1,1107 @@
+            graceful_shutdown_timeout: Amount of time after shutdown is called
+                that activities are given to complete before their tasks are
+                cancelled.
+            shared_state_manager: Used for obtaining cross-process friendly
+                synchronization primitives. This is required for non-async
+                activities where the activity_executor is not a
+                :py:class:`concurrent.futures.ThreadPoolExecutor`. Reuse of
+                these across workers is encouraged.
+            type_hint_eval_str: Whether the type hinting that is used to
+                determine dataclass parameters for decoding uses evaluation on
+                stringified annotations. This corresponds to the eval_str
+                parameter on :py:meth:`inspect.get_annotations`.
+        """
+        # TODO(cretz): Support workflows
+        if not activities:
+            raise ValueError("At least one activity must be specified")
+        # If there are any non-async activities, an executor is required
+        self._activities: Dict[str, _ActivityDefinition] = {}
+        for name, activity in activities.items():
+            if not callable(activity):
+                raise TypeError(f"Activity {name} is not callable")
+            elif not activity.__code__:
+                raise TypeError(f"Activity {name} does not have __code__")
+            elif activity.__code__.co_kwonlyargcount:
+                raise TypeError(f"Activity {name} cannot have keyword-only arguments")
+            elif not inspect.iscoroutinefunction(activity):
+                if not activity_executor:
+                    raise ValueError(
+                        f"Activity {name} is not async so an activity_executor must be present"
+                    )
+                if (
+                    not isinstance(
+                        activity_executor, concurrent.futures.ThreadPoolExecutor
+                    )
+                    and not shared_state_manager
+                ):
+                    raise ValueError(
+                        f"Activity {name} is not async and executor is not thread-pool executor, "
+                        "so a shared_state_manager must be present"
+                    )
+                if isinstance(
+                    activity_executor, concurrent.futures.ProcessPoolExecutor
+                ):
+                    # The function must be picklable for use in process
+                    # executors, we we perform this eager check to fail at
+                    # registration time
+                    # TODO(cretz): Is this too expensive/unnecessary?
+                    try:
+                        pickle.dumps(activity)
+                    except Exception as err:
+                        raise TypeError(
+                            f"Activity {name} must be picklable when using a process executor"
+                        ) from err
```

Comment:
Do we really need this? If users want to spawn *within* their activities, they can always do that. Not sure we need to support this form out-of-the-box. If it's asked for a bunch, maybe we add it then. Process management always ends up being a huge pain.

Reply:
It's mostly harmless to support. They are still doing the process management, but it makes for a nice setup if you support common Python async (asyncio, threading, multiprocess) out of the box. This eager check is just to save them if they are using multiprocessing. If we had to do process management I'd agree we shouldn't support it, but in this file we just make sure our stuff works across process, not managing processes.